synapse/handlers/presence.py

```diff
@@ -1080,12 +1196,23 @@ async def update_external_syncs_row(
                 process_id, set()
             )
 
+            devices = self.user_to_device_to_current_state.setdefault(user_id, {})
+            device_state = devices.setdefault(
+                device_id, UserDevicePresenceState.default(user_id, device_id)
+            )
+            device_state.last_sync_ts = sync_time_msec
+
             updates = []
-            if is_syncing and user_id not in process_presence:
-                if prev_state.state == PresenceState.OFFLINE:
+            if is_syncing and (user_id, device_id) not in process_presence:
+                if device_state.state == PresenceState.OFFLINE:
+                    # Mark the device as online since it is not syncing, this
+                    # might mark the user as online, so send an update.
+                    device_state.state = PresenceState.ONLINE
+                    presence = _combine_device_states(devices.values())
```

Comment:
I feel like a lot of this logic is a bit scattered around? I'd expect the state machine conditions to be in one place where we update the device info?

Reply:
Yes, I did not love how this is scattered around. Will give it a think! 👍

Reply:
I think this is pretty simplified now, but we still do call `_combine_presence_state` in a few spots.